Users crud

Andrevictor20 opened the pull request:
Foi adicionado 3 novas rotas para a realização do CRUD do usuário: create-users (para criar um novo usuário no banco de dados), get-users (para buscar um usuário existente no banco de dados), delete-users ( para remover um usuário do banco de dados). A rota create-users possui uma verificação para saber se um usuário já existe no banco de dadose se não houver cria o usuário e retorna a mensagem de confirmação, a rota get-users faz a busca do usuário pelo ID e retorna uma mensagem contendo o nome e cpf do usuário e a rota delete-users também faz a busca do usuário pelo ID e retorna uma mensagem de confirmação do usuário deletado com as informações do ID e nome dele. Todas as rotas possuem tratamento de erro, no caso da rota delete-users se um usuário já foi deletado ele irá exibir "Record to delete does not exist.". Além disso foram feitas alterações no schema.prisma e consequentemente na migration do create table user, as mudanças foram : Alterções no tipo do id e do cpf que passaram a ser String , pois anteriormente haviam conflitos ao usar as rotas em relação à tipagem ( ocasionada pelo Zod)

Andrevictor20 committed:
feat: Criação da tabela de usuários no schema.prisma, criação da rota createUsers, para criar um novo usuário

Andrevictor20 committed:
feat: Criação das rotas get-users (para buscar usuários), delete-users (para apagar usuários), registro dessas rotas no server.ts, alterções no model do user ( alterções no tipo do id, de autoincrement e inteiro para String) no schema.prisma

Andrevictor20 committed:
feat: Adição de verificação de existência de um usuário na rota create-users, tratamento de erros nas rotas create-users,get-users e delete-users, alterações no schema.prisma no model User, no tipo do id e no cpf, para evitar erros.

gabrielcancio commented on prisma/schema.prisma:
```diff
@@ -11,9 +12,19 @@ datasource db {
 
-}
\ No newline at end of file
+  @@map("user")
```
Nome da tabela deve estar no plural. Seria "users " no caso.

gabrielcancio commented on prisma/schema.prisma:
```diff
@@ -11,9 +12,19 @@ datasource db {
+  name      String
+  cpf       String  @unique
+  email     String   @unique
+  createdAt DateTime @default(now())
```
Os campos da tabela no campo devem ser apenas snake_case e o camelCase apenas para objetos JS/TS. Para fazer isso é só usar a função `@map` nos campos conforme a documentação do Prisma: https://www.prisma.io/docs/orm/prisma-schema/data-model/models#mapping-model-names-to-tables-or-collections

gabrielcancio commented on prisma/schema.prisma:
```diff
@@ -2,7 +2,8 @@
 generator client {
-  provider = "prisma-client-js"
+  provider      = "prisma-client-js"
+  binaryTargets = "debian-openssl-3.0.x"
```
Qual a finalidade desse binaryTargets? Ocasionou erro ao tentar o docker no meu ambiente local

Andrevictor20 replied:
Ocasionou um erro em minha máquina (Linux) ao tentar executar o código, e o próprio prisma recomendou adicionar o binaryTargets e salvar, dessa forma consegui executar na minha máquina.

gabrielcancio commented on prisma/schema.prisma:
```diff
@@ -11,9 +12,19 @@ datasource db {
 }
 
 model TabelaTeste {
```
Remover esta tabela de teste e toda a rota referente a teste. Ela só servia para testar se o servidor e a conexão com o banco.

Andrevictor20 replied:
OK.

gabrielcancio commented on src/infra/http/routes/create-users.ts:
```diff
@@ -0,0 +1,47 @@
+      if (existingUser) {
+        return reply.status(400).send({
+          message: 'User already exists',
+          user:{
```
Creio que não seria necessário devolver o usuário. Imagine o cenário em que um usuário malicioso esteja tentando descobrir o id de usuário específico. Dessa forma estaríamos entregando de bandeja. Creio que só a mensagem é suficiente

Andrevictor20 replied:
Irei remover, estava apenas testando se ele devolveria o usuário corretamente.

gabrielcancio commented on src/infra/http/routes/create-users.ts:
```diff
@@ -0,0 +1,47 @@
+      return reply.status(201).send({ message: 'User created!', user: user });
+
+    } catch (error) {
+      return reply.status(500).send({
```
Precisa melhorar esse tratamento de erro, desse jeito os erros gerados pelo zod na validação dos campos da request seriam retornados como erro 500 Internal Server Error, sendo que um erro dessa natureza seria um 400 Bad Request. 

Erros 500 Internal Server Error, são erros inesperados ocasionados pelo mal funcionamento de uma biblioteca ou de código, problemas de conexão com o banco ou qualquer coisa que estejam foram do nosso  controle ou expectativa.

Erros 400 Bad Request são erros relacionados a request que impedem a mesma de ser finalizada com sucesso, como por exemplo inputs indevidos, neste caso o usuário já existia e por tanto não foi concluída com sucesso. Faltar um input necessário também configura Bad Request.

gabrielcancio commented on src/infra/http/routes/delete-user.ts:
```diff
@@ -0,0 +1,32 @@
+        }
+      })
+
+      return reply.status(201).send({
```
HTTP Status code 201 Created é para justamente uma requisão que obteve sucesso ao efetuar uma requisição que cria algum recurso no servidor. Outros métodos seriam mais adequados como 200 OK, 202 Accepted ou até mesmo 204 No Content se não houver conteúdo na response. Link para doc falando sobre delete requests: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE e Link para doc de HTTP Status https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

gabrielcancio commented on src/infra/http/routes/get-users.ts:
```diff
@@ -0,0 +1,30 @@
+
+    })
+    if(user){
+      return reply.status(201).send({ user: user});
```
Aqui o HTTP Status Code correto deveria ser 200 OK

gabrielcancio commented on src/infra/http/server.ts:
```diff
@@ -1,8 +1,15 @@
+import { deleteUsers } from "./routes/delete-user";
 const app = fastify();
 app.register(teste)
+app.register(createUsers)
```
Está faltando a rota de alteração de dados do usuário.

gabrielcancio commented on src/infra/http/server.ts:
```diff
@@ -1,8 +1,15 @@
+import { getUsers } from "./routes/get-users";
+import { deleteUsers } from "./routes/delete-user";
 const app = fastify();
 app.register(teste)
```
Remover rota de teste

Andrevictor20 committed:
feat: Criação da rota update-users, para atualizar um usuário que é buscado pelo id, caso ele exista, caso contrário será retornado notFound

gabrielcancio commented on src/infra/http/routes/get-users.ts:
```diff
@@ -0,0 +1,30 @@
+  app.get('/user/:userId', async (request, reply) => {
+   try {
+    const getUserParams = z.object({
+      userId: z.string().uuid(),
```
Rota de get-users não está funcionando. Na validação do query param da request ele ta tentando validar um uuid sendo que o campo id no banco é inteiro.

gabrielcancio commented on src/infra/http/routes/update-users.ts:
```diff
@@ -0,0 +1,58 @@
+  app.put('/user/:userId', async (request, reply) => {
+    try {
+      const updateUserSchema = z.object({
+        userId: z.string().uuid(),
```
Validação por uuid falhando devido ao userId ser um integer

gabrielcancio commented on src/infra/http/routes/delete-user.ts:
```diff
@@ -0,0 +1,32 @@
+      });
+
+    } catch (error) {
+      return reply.status(500).send({ message: error.meta.cause})
```
error.meta é undefined, não está devolvendo nem o erro correto. Pelo visto não existe um padrão para o json da reponse, em alguns tem messaga outros tem error, necessário definir um padrão. Substituir onde houver error.meta no código.

gabrielcancio commented on src/infra/http/routes/update-users.ts:
```diff
@@ -0,0 +1,58 @@
+        email: z.string().email(),
+      })
+      const { userId } = updateUserSchema.parse(request.params);
+      const { name, email } = updateDataUserSchema.parse(request.body);
```
Desse jeito a request vai falhar se o usuário quiser alterar apenas um dos, os campos tem que ser opcionais.

Andrevictor20 replied:
Acho melhor na verdade alterar apenas o email, dificilmente alguém alteraria o nome, vou ajustar também.

gabrielcancio replied:
Rapaz supondo que a pessoa errou o nome na hora de se cadastrar e não percebeu(Facilmente eu faria isso), a pessoa vai ter que deletar a conta se quiser colocar o nome correto kkkk. Creio é valido manter a alteração do nome.

gabrielcancio reviewed (changes requested):
Testei as rotas e nenhuma ta funcionando. Necessário rever os pontos comentados e corrigir o código.

Andrevictor20 committed:
fix: Correção dos status code no tratamento de erros de todas as rotas, remoção da rota e da tabela teste, alterções de nomenclatura na tabela users, alterção da PORT do postgree do docker de 5432 para 5433, correções no update.user para funcionar apenas mudando o email

gabrielcancio commented on docker-compose.yaml:
```diff
@@ -18,7 +18,7 @@ services:
     image: postgres:16.2-alpine
     ports:
-      - '5432:5432'
+      - '5433:5433'
```
Commitou as alterações de porta do docker, isso é apenas na sua máquina. O bind de portas está errado, ta fazendo o bind na porta 5433 da tua máquina e na porta 5433 do container, mas dentro do container o postegres roda na 5432, se tu tentar usar um cliente externo de postgres ele não vai funcionar (Exceto o Prisma Studio).

gabrielcancio commented on prisma/schema.prisma:
```diff
@@ -2,18 +2,22 @@
+  name      String
+  cpf       String   @unique
+  email     String   @unique
+  createdat DateTime @default(now())
```
Aqui continua errado kkk. O padrão para as nomes de campos e tabelas no banco é snake_case, ou seja tudo em minusculo e mais de uma palavra separada por vírgula, como por exemplo `createdAt`, ficaria `created_at`. Entretanto no javascript deveria continuar camelCase, ou seja createdAt continuaria desse mesmo jeito dentro do javascript.

Mas como fazer isso?
Bom eu falei que de modo semelhante ao nome da tabela, da para usar o `@map` para renomear apenas o nome da coluna no banco e continuar do jeito declarado no javascript. Mais uma vez, segue o link para doc: https://www.prisma.io/docs/orm/reference/prisma-schema-reference#map

gabrielcancio commented on src/infra/http/routes/create-users.ts:
```diff
@@ -0,0 +1,43 @@
+      return reply.status(201).send({ message: 'User created!', user: user });
+
+    } catch (error) {
+      return reply.status(400).send({
```
Com relação ao tratamento de erros, ou falei que deveria ser melhor. No caso queria mais especificidade, poder destinguir erros internos de bad request e assim por diante, desse modo tu só tirou o erro 500 e substituiu pelo 400. É preciso ser mais específico. Esse deixa para a gente fazer em pair (Call).

gabrielcancio commented on src/infra/http/routes/get-users.ts:
```diff
@@ -0,0 +1,30 @@
+import { FastifyInstance } from "fastify";
+import { prisma } from '../../client/prisma'
+import {z} from 'zod'
+export async function getUsers(app: FastifyInstance) {
```
Creio que o nome getUser ficaria melhor, pois tu está retornando apenas os dados de um usuário em específico. getUsers da impressão que tu está listando todos os usuários.

Escolher nomes corretos é essencial. O nome precisa em pouquíssimas palavras dar uma clara idéia do que aquele código faz. 

Código legível é um dos pontos que separam os bons programadores de maus programadores porque em uma empresa você não está fazendo código para você, está fazendo para os outros. Quando outra pessoa, ou até mesmo você voltar nesse código no futuro não vai se lembrar 100% do contexto. Nomes ambíguos vão confundir e fazer que você tenha mais trabalho para entender o funcionamento do código para melhorá-lo ou corrigi-lo dimunindo a produtividade do time em geral.

gabrielcancio commented on src/infra/http/routes/get-users.ts:
```diff
@@ -0,0 +1,30 @@
+      }
+
+    })
+    if(user){
```
Recomendo evitar o uso de else o máximo possível, pois quantos mais elses associados a  ifs você tiver, aumenta um negócio chamado [complexidade cliclomática ](https://www.treinaweb.com.br/blog/complexidade-ciclomatica-analise-estatica-e-refatoracao), que basicamente falando seria você criar um labirinto de tantos ifs e elses que você usa que dificulta muito o entendimento do código, por isso é bom evita-lo.

Um truque para se guiar em quando se trata de compexlidade ciclomática é que o código deve estar o mais próximo da margem quanto possível.

Dúvidas pergunte.

gabrielcancio commented on src/infra/http/routes/update-users.ts:
```diff
@@ -0,0 +1,55 @@
+          email
+        },
+      });
+      return reply.status(201).send({
```
Continua usando o HTTP Status Code 201 created, que deve ser usado apenas para quando um recurso é criado.

Trocar por 200 OK ou 202 Accepted.

gabrielcancio reviewed (changes requested):
Mais uma leva de correções para fazer.

Andrevictor20 committed:
fix: Alterções das portas do postgre do arquivo docker.compose,
alterções no schema.prisma para conseguir executar o prisma localmente e no container do docker, alterções nas migrations (maior legilibilidade), alterções nas nomenclauturas dos arquivos das rotas( maior legibilidade)

Andrevictor20 committed:
fix: Alterações nos tratamentos de erros de todas as rotas, com maior detalhamento de erros.

Commit:
fix: fixing error handler using fastify and exceptions in routes

Commit:
fix: changing the way to deploy the migrations in entrypoint.sh

Commit:
fix: removing console.log call in fastify error handler

Commit:
refactor: improving the handle of interal server errors to standart

Commit:
fix: adding bad request exception to create user route flow

Commit:
refactor: abstracting error handler to own class

Commit:
refactor: changing name of the filename of create-users.ts to create-user

Commit:
fix: fixing import
